From patchwork Mon Oct 15 16:57:41 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10642199
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id A527013AD
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:57:49 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 91F46299B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:57:49 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 82B45299C2; Mon, 15 Oct 2018 16:57:49 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 31F49299B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:57:49 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726790AbeJPAnu (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 15 Oct 2018 20:43:50 -0400
Received: from mail-oi1-f202.google.com ([209.85.167.202]:34890 "EHLO
        mail-oi1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726557AbeJPAnu (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 15 Oct 2018 20:43:50 -0400
Received: by mail-oi1-f202.google.com with SMTP id f62-v6so13761661oia.2
        for <kvm@vger.kernel.org>; Mon, 15 Oct 2018 09:57:47 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=26moTWRnrmOBUeFbHsNqIECyumqhtsRGVo6xen3irHY=;
        b=WJkLYzmC5enpS58Vk6M1rO+pmozeUwdUjpwMfvZ9SwpdGE2EPK1C+qnjuOldfNy6Wk
         p+Htudzr4atoqIM3VFgVF7ZH0nXIhROgXuD+6ta4O9WdgRcB5LCXqsqAPVH+x5aIY56a
         z1S5pwV1XFeVL2qsMHJxmNggnBlAumKXpKh6NTlPuG4KmsO0/EzDP+jy4Fs3cswj69b3
         bFYWZcRvNdA284kWq7rUbzEdmL2mXUsbhdzlRcDzxIDA6yIBD5SrNFwKTfdIxpKjmD7O
         Imqk+oBd9T+dPv5H/bQUD/Q8iTjSVx2nwRjKM5QsZUrNS3aecsMxI6tQbdynoNEuYD+G
         zAFw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=26moTWRnrmOBUeFbHsNqIECyumqhtsRGVo6xen3irHY=;
        b=TxBL9U7v0VYhWX4w+aklhesBu47GY/XZcuAcZgQWSUmdStIqLjsjRlunmr2LUyk715
         rBtzdcwZ9IqiC+CDYkU4l928+7bO7VkvHFxcJ4dPYyVOCGgnojBYvpB1AppOvvGzgt/A
         h/0J09CvKgH4e6yepwV6mYOvUfKuf+3y8eL/0pThh8sQvqyw4y1haPEGbklzAvBA9T1r
         vPYtze9X/67AJdf3ehtPlMpGMfy+x9PW7x3HVFZ6hb/BPpHbIvVhueJkOTK+uojya2gC
         QARsqViUr6owav9yYKPUcvKzlO4P6qqudFJaUdvIbuahUxOOYJKd0am+/Q+MFR/0ttn1
         Oy6A==
X-Gm-Message-State: ABuFfojHyXPiaF7q+XSyVEU5k8DmX0SnhnTxS4och/hEL3AvxxcgQROM
        VHuh1GE0JQPD2Ucmc/87zWbTyG7I6Z0VMq1EUeST8nUPF4W50JAtOfgE67xUtxDv3sBPGVYVFNg
        Qc/L/0AemrECwMOl6SyMHjLtPr2XUoukMARijKBnccbcXT13i/X1qqdq8B+FjHv4=
X-Google-Smtp-Source: 
 ACcGV61GU5YdW4SrYMySakCq0OZ6bET8yOOknURPpDTOluvaOtrFHkn0y9bD+cactPej13zEgAIM0/46gT4pkQ==
X-Received: by 2002:a9d:55b6:: with SMTP id m51mr14793979oth.27.1539622666870;
 Mon, 15 Oct 2018 09:57:46 -0700 (PDT)
Date: Mon, 15 Oct 2018 09:57:41 -0700
In-Reply-To: <20181015164817.GA25227@linux.intel.com>
Message-Id: <20181015165742.38393-1-jmattson@google.com>
Mime-Version: 1.0
References: <20181015164817.GA25227@linux.intel.com>
X-Mailer: git-send-email 2.19.1.331.ge82ca0e54c-goog
Subject: [PATCH v3 1/2] kvm: x86: Bounds-check argument to
 x86_exception_has_error_code
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Marc Orr <marcorr@google.com>,
        Sean Christopherson <sean.j.christopherson@intel.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

x86_exception_has_error_code should deterministically return false if
its operand is greater than 32.

Fixes: 0447378a4a793 ("kvm: vmx: Nested VM-entry prereqs for event inj.")
Cc: Marc Orr <marcorr@google.com>
Cc: Sean Christopherson <sean.j.christopherson@intel.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 arch/x86/kvm/x86.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/x86/kvm/x86.h b/arch/x86/kvm/x86.h
index 67b9568613f3..f6965f9734dc 100644
--- a/arch/x86/kvm/x86.h
+++ b/arch/x86/kvm/x86.h
@@ -116,6 +116,8 @@ static inline bool x86_exception_has_error_code(unsigned int vector)
 			BIT(NP_VECTOR) | BIT(SS_VECTOR) | BIT(GP_VECTOR) |
 			BIT(PF_VECTOR) | BIT(AC_VECTOR);
 
+	if (WARN_ON_ONCE(vector >= 32))
+		return false;
 	return (1U << vector) & exception_has_error_code;
 }
 

From patchwork Mon Oct 15 16:57:42 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10642201
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7E89013AD
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:58:01 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6E701299B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:58:01 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 6225A299C2; Mon, 15 Oct 2018 16:58:01 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0CAA3299B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:58:01 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726827AbeJPAoC (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 15 Oct 2018 20:44:02 -0400
Received: from mail-pl1-f201.google.com ([209.85.214.201]:47471 "EHLO
        mail-pl1-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726791AbeJPAoC (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 15 Oct 2018 20:44:02 -0400
Received: by mail-pl1-f201.google.com with SMTP id t10-v6so553264plh.14
        for <kvm@vger.kernel.org>; Mon, 15 Oct 2018 09:57:59 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=gH3vOpfOfjsd5JATN5AuGpFIYfOc4r8pj+3P1JHlAwM=;
        b=XKZVTxjrjZ70Uma2jzS6ZItzsY7Pf3f08tUMPPFVa71fPDRjO9sYk8LRMQSgUvaFLd
         vYJCIf8TCWKuywbadnewdkBrzKIGRpH9Oro8YrzrDyfrgjSWe+ZzVb+OvPux7S209KeL
         kEGM69YBXluk9uhwxM5jpa/92Zin3SsH4rpqPrO0GlKkVA/fzKQ5SveK/3h3cgw8RcYQ
         IVnlAdAAhNpOyR6a8C0bwzr6AaIuB2Tpy6LIpr0JsSizBhuLowPvEsfhzTsDXhrNY4ew
         MZ17uweWVNqLdH/WMiTVUTNPzl7l1N4toygfoASt8EMtePDZ9bfs33umCI7Zlb/fBzh4
         xSqw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=gH3vOpfOfjsd5JATN5AuGpFIYfOc4r8pj+3P1JHlAwM=;
        b=P68kNOgiCGg5yB+PmJTVXUGvTU/TTzPkirKPTQjrvBAuVmEfGe4smsFbq4Gg5OLLKp
         yXUSvsz04YS6FnFWataUA7RPUhEyTtdDlbOplLkTbYvmTSzlzOOWRyICRlCN/ARIoxjq
         VziJYVJQAzeWd0N8CUmyBGFSyOtDGDQls7DGvJV8nHzkdOdaS+/wZ+e6GiokdAN1dasC
         /DuJdIJ/87i88GvIMWkmdwA6NOPEl/bWHNbtQ5rlAfqBKcFoSzHGnnSzbRssVDGY37V1
         oDfv2kWJLjQMVt054HljnuZjCvjgjIkMUcuoaGR8oCqTkPKPZC4t1SumB6PH6TQSgj7K
         7geg==
X-Gm-Message-State: ABuFfogHYTQ9VYulT5JPRYPTmJSF20ScsLEmCjg1iKwQTV2ffdHQQOaZ
        60YgZB6BTI2Ub+En2kTf9Vkn86fBffK6SfIgFVaAOlqphd9gxQn4QQt+/8Qj7lfTExMeyCI+wM4
        Q/ahc9OC8FS+wglNCVemvDqsUEFZJMNeHITqg73smsVNEb1y7pMR6pvGcvKFeKqM=
X-Google-Smtp-Source: 
 ACcGV61TLB1Rv5I78CdeF1NJNzVIGhHM+SRUu4O0b2pdVvBOmPV4qSdx1tXBjQinztGn/f/Bin+hkkzFq0nPSQ==
X-Received: by 2002:a62:c91d:: with SMTP id
 k29-v6mr8976732pfg.25.1539622678970;
 Mon, 15 Oct 2018 09:57:58 -0700 (PDT)
Date: Mon, 15 Oct 2018 09:57:42 -0700
In-Reply-To: <20181015165742.38393-1-jmattson@google.com>
Message-Id: <20181015165742.38393-2-jmattson@google.com>
Mime-Version: 1.0
References: <20181015164817.GA25227@linux.intel.com>
 <20181015165742.38393-1-jmattson@google.com>
X-Mailer: git-send-email 2.19.1.331.ge82ca0e54c-goog
Subject: [PATCH v3 2/2] kvm: x86: Disallow invalid exceptions from userspace
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Userspace should not be able to inject an invalid exception into a kvm
guest via KVM_SET_VCPU_EVENTS, nor should it be able to convert a
valid pending exception to an invalid one via KVM_SET_SREGS.

In particular, only certain exceptions deliver an error code in
protected mode, and no exception delivers an error code in
real-address mode.

Reported-by: syzbot <syzkaller@googlegroups.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
---
 arch/x86/kvm/x86.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 4ab569171ad1..b67a664b1d7b 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -3423,6 +3423,14 @@ static int kvm_vcpu_ioctl_x86_set_vcpu_events(struct kvm_vcpu *vcpu,
 	    vcpu->arch.mp_state == KVM_MP_STATE_INIT_RECEIVED)
 		return -EINVAL;
 
+	if (events->exception.injected) {
+		bool has_error_code = is_protmode(vcpu) &&
+			x86_exception_has_error_code(events->exception.nr);
+
+		if (!!events->exception.has_error_code != has_error_code)
+			return -EINVAL;
+	}
+
 	process_nmi(vcpu);
 	vcpu->arch.exception.injected = false;
 	vcpu->arch.exception.pending = events->exception.injected;
@@ -8170,6 +8178,14 @@ static int kvm_valid_sregs(struct kvm_vcpu *vcpu, struct kvm_sregs *sregs)
 			return -EINVAL;
 	}
 
+	if (vcpu->arch.exception.injected || vcpu->arch.exception.pending) {
+		bool has_error_code = (sregs->cr0 & X86_CR0_PE) &&
+			x86_exception_has_error_code(vcpu->arch.exception.nr);
+
+		if (vcpu->arch.exception.has_error_code != has_error_code)
+			return -EINVAL;
+	}
+
 	return 0;
 }
 
